GoogleUtilities/Environment/third_party/GULHeartbeatDateStorage.m

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#import "GULHeartbeatDateStorage.h"
+#import <GoogleUtilities/GULLogger.h>
+#import <GoogleUtilities/GULSecureCoding.h>
+
+static GULLoggerService kGULHeartbeatDateStorage = @"GULHeartbeatDateStorage";
+
+@interface GULHeartbeatDateStorage ()
+/** The storage to store the date of the last sent heartbeat. */
+@property(nonatomic, readonly) NSFileCoordinator *fileCoordinator;
+@end
+
+@implementation GULHeartbeatDateStorage
+
+- (instancetype)initWithFileName:(NSString *)fileName {
+  if (fileName == nil) {
+    return nil;
+  }
+
+  self = [super init];
+  if (self) {
+    _fileCoordinator = [[NSFileCoordinator alloc] initWithFilePresenter:nil];
+    NSURL *directoryURL = [[self class] directoryPathURL];
+    [[self class] checkAndCreateDirectory:directoryURL:_fileCoordinator];
+    _fileURL = [directoryURL URLByAppendingPathComponent:fileName];
+  }
+  return self;
+}
+
+/** Returns the URL path of the Application Support folder.
+ * @return the URL path of Application Support.
+ */
++ (NSURL *)directoryPathURL {
+  NSArray<NSString *> *paths =
+      NSSearchPathForDirectoriesInDomains(NSApplicationSupportDirectory, NSUserDomainMask, YES);
+  NSArray<NSString *> *components = @[ paths.lastObject, @"Google/FIRApp" ];
+  NSString *directoryString = [NSString pathWithComponents:components];
+  NSURL *directoryURL = [NSURL fileURLWithPath:directoryString];
+  return directoryURL;
+}
+
+/** Checks and creates a directory for the directory specified by the
+ * directory url
+ * @param directoryPathURL The path to the directory which needs to be created.
+ * @param fileCoordinator The fileCoordinator object to coordinate writes to the directory.
+ */
++ (void)checkAndCreateDirectory:(NSURL *)directoryPathURL:(NSFileCoordinator *)fileCoordinator {
```

Comment:
nit: It should be `+ (void)checkAndCreateDirectory:(NSURL *)directoryPathURL fileCoordinator:(NSFileCoordinator *)fileCoordinator`. Can be in a followup PR as well.